### PR TITLE
feat(query): table function generate_series & range

### PR DESCRIPTION
**Pull request by jun0315**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
generate_series can be a table function.

support Number, Timestamp and Date argument type.

`range` is special case for `generate_series`, it's similar to python's range function.
Yet `generate_series` is inclusive_bound and `range` is exclusive_bound.

Tracking https://github.com/datafuselabs/databend/issues/10400

### CODE REVIEW

File: src/query/service/src/table_functions/srf/generate_series.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+            Ok(Some(DataBlock::new_from_columns(vec![
+                Int64Type::from_data(range(start, end, step).into_iter()),
+            ])))
+        }
+    }
+}
+
+pub fn number_scalars_have_float(scalars: Vec<Scalar>) -> bool {
+    for scalar in scalars {
+        match scalar {
```
Comment (sundy-li) [inline src/query/service/src/table_functions/srf/generate_series.rs:363]: `matches!(...)`

File: src/query/service/src/table_functions/srf/generate_series.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        if self.finished {
+            return Ok(None);
+        }
+        let have_float = number_scalars_have_float(vec![
+            self.start.clone(),
+            self.end.clone(),
+            self.step.clone(),
+        ]);
+        if have_float {
+            let dest_type = DataType::Number(NumberDataType::Float64);
```
Comment (sundy-li) [inline src/query/service/src/table_functions/srf/generate_series.rs:255]: 1. better to have a `macro` to make the codes look better.
2. `date`, `timestamp` should be covered.

File: src/query/service/src/table_functions/srf/generate_series.rs (hunk 1 above)
Comment (RinChanNOWWW) [reply src/query/service/src/table_functions/srf/generate_series.rs:363]: How about add a method `is_float` for `Scalar`.

And check `if arg1.is_float() || arg2.is_float() || arg3.is_float()`.

File: src/query/service/src/table_functions/srf/generate_series.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+                NumberScalar::UInt64(_) => (),
+                NumberScalar::Int8(_) => (),
+                NumberScalar::Int16(_) => (),
+                NumberScalar::Int32(_) => (),
+                NumberScalar::Int64(_) => (),
+                NumberScalar::Float32(_) => return true,
+                NumberScalar::Float64(_) => return true,
+            },
+            Scalar::Decimal(_) => return true,
+            _ => (),
```
Comment (RinChanNOWWW) [inline src/query/service/src/table_functions/srf/generate_series.rs:377]:
```suggestion
        match scalar {
            Scalar::Number(NumberScalar::Float32(_)) | Scalar::Number(NumberScalar::Float32(_)) |  Scalar::Decimal(_) =>  return true,
            _ => {},
```

File: src/query/service/src/table_functions/srf/generate_series.rs [hunk 4]
```diff
@@ -379,3 +358,27 @@ pub fn number_scalars_have_float(scalars: Vec<Scalar>) -> bool {
...
+        table_args.positioned.len(),
+        Some((2, 3)),
+        2,
+    )?;
+
+    if args
+        .iter()
+        .all(|arg| matches!(Scalar::Number(_) | Scalar::Decimal(_)))
+        || args.iter().all(|arg| matches!(Scalar::Date(_)))
+        || args.iter().all(|arg| matches!(Scalar::Timestamp(_)))
```
Comment (RinChanNOWWW) [inline src/query/service/src/table_functions/srf/generate_series.rs:375]:
```suggestion
    if args
        .iter()
        .all(|arg| matches!(Scalar::Number(_) | Scalar::Decimal(_) | Scalar::Date(_) | Scalar::Timestamp(_))
```

**Pull request merged**